File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/javascript/reverse_string.js
function reverseString(str) {
  const splitString = str.split('');
  let reverseString = '';
  splitString.forEach(element => reverseString = element + reverseString);
  return reverseString
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));
}

module.exports = reverseString;

//psuedocode:

// function reverseString(str) 
//   split String
//   create empty String
//   iterate over split string 
//   add element to beginning of empty string creating reverse string
//   return reverse string
// done

//explanation:

// I split the string to be able to work on each element of string
// I then iterate on the split string which is now an Array and add each element
// to the beginning of an empty string 
// by doing so creating a reverse string

File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/ruby/reverse_string.rb
def reverse_string(str)
  splitString = str.split('')
  reverseString = ''
  splitString.each{|element| reverseString.prepend(element)}
  reverseString
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  # Don't forget to add your own!
  puts "Expecting: 'repmac yppah'"
  puts "=>", reverse_string('happy camper')

end

# psuedocode:

# reversestring(arg)
#  create a split string from arg = [a r g]
#  create and empty string = ''
#  iterate over split string and prepend each element to empty string creating reverse string
#  return reverse string
# end

# Description:

# In this example I am breaking apart the string so I can work on each element of the string individually
# I then prepend each element to the new string
# another way to do this is create a new array and append each element to the new array then join array together
# I am curious which is more readable and faster
# then return the string
